File: leetcode/leetcode_py/determine-if-two-strings-are-close.py
'''
https://leetcode.com/problems/determine-if-two-strings-are-close

1. list and sort
time: o(nlogn) n(word size) + nlogn (26log26)
Swap or Replace all occurence
has to meet
1. same frequency (no matter chars)
ex) a:4, b:1 == b:4, a:1
2. same set values (that can replace or swap)
ex) a,b == a,b
'''
class Solution:
    def closeStrings_1(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2): return False
        c1 = [0] * 26
        c2 = [0] * 26
        for i in range(len(word1)):
            c1[ord(word1[i])-ord('a')] += 1
            c2[ord(word2[i])-ord('a')] += 1
        for i in range(26):
            if (c1[i] and not c2[i]) or (not c1[i] and c2[i]):
                return False
        c1.sort()
        c2.sort()
        return c1 == c2

    def closeStrings_2(self, word1: str, word2: str) -> bool:
        s1 = set(word1)
        s2 = set(word2)
        if s1 != s2:
            return False
        c1 = sorted(Counter(word1).values())
        c2 = sorted(Counter(word2).values())
        return c1 == c2

    def closeStrings(self, word1: str, word2: str) -> bool:
        return self.closeStrings_2(word1, word2)

File: leetcode/leetcode_cpp/determine-if-two-strings-are-close.cpp
/*
https://leetcode.com/problems/determine-if-two-strings-are-close

1. list and sort
time: o(nlogn) n(word size) + nlogn (26log26)
space: o(n)
Swap or Replace all occurence conditions
1. same set values (to enable swap, the char exist on each other)
ex) a,b == a,b
2. same frequency (no matter chars, to replace)
ex) a:4, b:1 == b:4, a:1
*/
class Solution {
public:
    bool closeStrings(string word1, string word2) {
        if (word1.size() != word2.size()) return false;
        vector<int> c1(26, 0);
        vector<int> c2(26, 0);
        for (int i=0; i<word1.size(); ++i) {
            c1[word1[i] - 'a']++;
            c2[word2[i] - 'a']++;
        }
        for (int i=0; i<26; ++i) {
            if ((c1[i] && !c2[i]) || (!c1[i] && c2[i])) return false;
        }
        sort(begin(c1), end(c1));
        sort(begin(c2), end(c2));
        return c1 == c2;
    }
};